topdown/eval.go

```diff
@@ -2378,6 +2386,17 @@ func (e evalVirtualPartial) evalEachRule(iter unifyIterator, unknown bool) error
 	}
 
 	result := e.empty
+	if e.ir.DynamicRef {
+		for _, rule := range e.ir.Rules {
+			result, err = e.evalOneDynamicRefRulePreUnify(rule, result, unknown)
+			if err != nil {
+				return err
+			}
+		}
+		e.e.virtualCache.Put(hint.key, result)
+		return e.evalTerm(iter, e.pos+1, result, e.bindings)
+	}
+
 	for _, rule := range e.ir.Rules {
```

Comment:
Given that generic ref head rules are the more general case, you might expect that the base case (below) is also covered by what's added above -- is that so? Or what am I missing?

Reply:
I think you're right. These two branches should be possible to merge into one; given that there is no performance benefit to keeping them separate.

I think it's around here PE is breaking down too; so I'll dig around here a bit to see what improvements are necessary + possible.